component/src/main/java/org/ballerinalang/net/jms/actions/Poll.java

```diff
@@ -0,0 +1,127 @@
+import org.ballerinalang.connector.api.ConnectorUtils;
+import org.ballerinalang.model.types.TypeKind;
+import org.ballerinalang.model.values.BConnector;
+import org.ballerinalang.model.values.BStruct;
+import org.ballerinalang.nativeimpl.actions.ClientConnectorFuture;
+import org.ballerinalang.natives.annotations.Argument;
+import org.ballerinalang.natives.annotations.BallerinaAction;
+import org.ballerinalang.natives.annotations.ReturnType;
+import org.ballerinalang.net.jms.Constants;
+import org.ballerinalang.net.jms.JMSUtils;
+import org.ballerinalang.util.exceptions.BallerinaException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.transport.jms.contract.JMSClientConnector;
+import org.wso2.transport.jms.exception.JMSConnectorException;
+import org.wso2.transport.jms.impl.JMSConnectorFactoryImpl;
+import org.wso2.transport.jms.sender.wrappers.SessionWrapper;
+import org.wso2.transport.jms.utils.JMSConstants;
+
+import java.util.Map;
+import java.util.UUID;
+import javax.jms.Message;
+
+import static org.ballerinalang.net.jms.Constants.EMPTY_CONNECTOR_ID;
+
+/**
+ * {@code Poll} is the poll action implementation of the JMS Client Connector.
+ */
+@BallerinaAction(packageName = "ballerina.net.jms",
+                 actionName = "poll",
+                 connectorName = Constants.CONNECTOR_NAME,
+                 args = {
+                         @Argument(name = "queueName",
+                                   type = TypeKind.STRING),
+                         @Argument(name = "timeout", type = TypeKind.INT)
+                 },
+                 returnType = {@ReturnType(type = TypeKind.STRUCT, elementType = TypeKind.STRUCT,
+                                           structPackage = "ballerina.net.jms", structType = "JMSMessage"),
+                               @ReturnType(type = TypeKind.STRUCT)})
+public class Poll extends AbstractJMSAction {
+    private static final Logger log = LoggerFactory.getLogger(Poll.class);
+
+    @Override
+    public ConnectorFuture execute(Context context) {
+
+        // Extract argument values
+        BConnector bConnector = (BConnector) getRefArgument(context, 0);
+        String destination = getStringArgument(context, 0);
+        int timeout = getIntArgument(context, 0);
+
+        validateParams(bConnector);
+
+        // Get the map of properties.
+        BStruct  connectorConfig = ((BStruct) bConnector.getRefField(0));
+
+        Map<String, String> propertyMap = JMSUtils.preProcessJmsConfig(connectorConfig);
+
+        // Generate connector the key, if its not already generated
+        String connectorKey;
+        if (EMPTY_CONNECTOR_ID.equals(bConnector.getStringField(0))) {
+            connectorKey = UUID.randomUUID().toString();
+            bConnector.setStringField(0, connectorKey);
+        } else {
+            connectorKey = bConnector.getStringField(0);
+        }
+
+        propertyMap.put(JMSConstants.PARAM_DESTINATION_NAME, destination);
```

Comment:
shall have a config object with known properties and the separate map for only unknown properties?

Reply:
this has to be provided from transport-jms, created an issue for that https://github.com/wso2/transport-jms/issues/9